packages/proxy/lib/http/response-middleware.ts

```diff
@@ -393,101 +393,105 @@ const MaybePreventCaching: ResponseMiddleware = function () {
+  //   automatically.
+  // - In the request middleware, we retrieve the cookies for a given URL
+  //   and attach them to the request, like the browser normally would.
+  //   tough-cookie handles retrieving the correct cookies based on domain,
+  //   path, etc. It also removes cookies from the cookie jar if they've expired.
+  const needsCrossOriginHandling = checkNeedsCrossOriginHandling(this)
+
+  const appendCookie = (cookie) => {
+    const headerName = needsCrossOriginHandling ? 'X-Set-Cookie' : 'Set-Cookie'
+
+    try {
+      this.res.append(headerName, cookie)
+    } catch (err) {
+      this.debug(`failed to append header ${headerName}, continuing %o`, { err, cookie })
     }
-  } else if (!isFirefox || url.protocol === 'https:') {
-    ctxDebug('add Secure to cookie')
-    cookie += '; Secure'
   }
 
-  ctxDebug('resulting cookie: %s', cookie)
+  if (!this.config.experimentalSessionAndOrigin) {
+    ([] as string[]).concat(cookies).forEach((cookie) => {
+      appendCookie(cookie)
+    })
 
-  return cookie
-}
+    return this.next()
+  }
 
-const CopyCookiesFromIncomingRes: ResponseMiddleware = function () {
-  const cookies: string | string[] | undefined = this.incomingRes.headers['set-cookie']
+  const cookiesHelper = new CookiesHelper({
+    cookieJar: this.getCookieJar(),
+    currentAUTUrl: this.getAUTUrl(),
+    debug: this.debug,
+    request: {
+      url: this.req.proxiedUrl,
+      isAUTFrame: this.req.isAUTFrame,
+      needsCrossOriginHandling,
+    },
+  })
 
-  if (cookies) {
-    const needsCrossOriginHandling = (
-      this.config.experimentalSessionAndOrigin
-      && determineIfNeedsCrossOriginHandling(this)
-    )
-    const browser = this.getCurrentBrowser() || { family: null }
-    const url = new URL(this.req.proxiedUrl)
-    const isLocalhost = uri.isLocalhost(url)
+  await cookiesHelper.capturePreviousCookies()
 
-    this.debug('force SameSite=None?', needsCrossOriginHandling)
+  ;([] as string[]).concat(cookies).forEach((cookie) => {
+    cookiesHelper.setCookie(cookie)
 
-    ;([] as string[]).concat(cookies).forEach((cookie) => {
-      if (needsCrossOriginHandling) {
-        cookie = ensureSameSiteNone({ cookie, browser, isLocalhost, url, ctxDebug: this.debug })
-      }
+    appendCookie(cookie)
+  })
 
-      try {
-        this.res.append('Set-Cookie', cookie)
-      } catch (err) {
-        this.debug('failed to Set-Cookie, continuing %o', { err, cookie })
-      }
-    })
+  const addedCookies = await cookiesHelper.getAddedCookies()
+
+  if (!needsCrossOriginHandling || !addedCookies.length) {
+    return this.next()
   }
 
-  this.next()
+  try {
+    await this.getAutomation().request('set:cookies', addedCookies, () => {})
+  } catch (err) {
+    this.debug('failed setting cookies via automation: %s', err.message)
+  } finally {
+    this.next()
+  }
 }
 
+// TODO: error if this.next() is called twice in same block if possible
```

Comment:
will this TODO be handled in this PR? or do we need to log a followup issue?

Reply:
Yes! Forgot about that one. Addressed in [`3c81b36` (#22320)](https://github.com/cypress-io/cypress/pull/22320/commits/3c81b364c0ec56bbe1c769e4cd417fff51a19214)